server/src/main/java/io/crate/replication/logical/LogicalReplicationService.java

```diff
@@ -21,32 +21,97 @@
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreClusterStateListener;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
 import org.elasticsearch.cluster.ClusterChangedEvent;
 import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.repositories.RepositoriesService;
+import org.elasticsearch.snapshots.RestoreInfo;
+import org.elasticsearch.snapshots.RestoreService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.NoSuchRemoteClusterException;
+import org.elasticsearch.transport.RemoteClusterAware;
+import org.elasticsearch.transport.RemoteClusterAwareClient;
+import org.elasticsearch.transport.RemoteClusterConnection;
+import org.elasticsearch.transport.RemoteConnectionStrategy;
+import org.elasticsearch.transport.Transport;
+import org.elasticsearch.transport.TransportService;
 
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.io.IOException;
 import java.util.Collections;
+import java.util.Locale;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
 
-public class LogicalReplicationService implements ClusterStateListener {
+import static io.crate.replication.logical.repository.LogicalReplicationRepository.REMOTE_REPOSITORY_PREFIX;
+import static io.crate.replication.logical.repository.LogicalReplicationRepository.TYPE;
+import static org.elasticsearch.action.support.master.MasterNodeRequest.DEFAULT_MASTER_NODE_TIMEOUT;
 
+public class LogicalReplicationService extends RemoteClusterAware implements ClusterStateListener, Closeable {
+
+    private static final Logger LOGGER = LogManager.getLogger(LogicalReplicationService.class);
+
+    private final TransportService transportService;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private RepositoriesService repositoriesService;
+    private RestoreService restoreService;
+
+    private final Map<String, RemoteClusterConnection> remoteClusters = ConcurrentCollections.newConcurrentMap();
+    private final Map<String, String> subscribedIndices = ConcurrentCollections.newConcurrentMap();
+    /**
+     * Both vars are holding references to the latest/current metadata of the cluster state
+     */
     private SubscriptionsMetadata subscriptionsMetadata;
     private PublicationsMetadata publicationsMetadata;
```

Comment:
Shouldn't these also be volatile ? Afaik `clusterChanged` is called from different threads than some of the other methods accessing these.

(also, seeing the comment now - maybe renaming them to `currentSubscriptions` and `currentPublications` would make the comment unnecessary and convey the same information. Sorry, could've suggested that to begin with)


also - if we mark them as volatile I think we could initialize them to empty maps and ensure in the update paths that they never become null - to avoid duplicate volatile reads in the getter.